### ISSUE TITLE
Hello Lyberta. I think we have some common goals about IO. Can we have a discussion about it?

### ISSUE THREAD

**Issue by expnkx** [Lyberta/cpp-io#1]

**Comment by Lyberta** [comment]
Sure, you could have written your thoughts in the body of issue.

**Comment by expnkx** [comment]
> Sure, you could have written your thoughts in the body of issue.

With my project advancing, for me, the problem of current IO is worse than I think.

Have you ever tested the performance of fprintf, fout vs fast_io::obuf? The performance gap is insane. I think my project can show you how bad it is.

Do you have the latest GCC compiler? Here is my compilation of MinGW GCC on bitbucket.
https://bitbucket.org/ejsvifq_mabmip/mingw-gcc/src/master/

**Comment by Lyberta** [comment]
I agree that memory, file and socket IO should have the same design. At least on the high level that I'm working on.

You can also voice your opinion on [experimental IO study group](https://github.com/experimental-io). I've been in contact with Niall Douglas so he could review my proposal and offer advice. I want to present this proposal in Prague with him championing it. So far he's good with that.

I've tried to contact the main person behind Networking TS but couldn't yet.

Unicode is slightly different. Without going to files and sockets you don't really need to care about endianness so Unicode layer can sit on top of byte layer and sometimes even don't use it. That will require a different design. [I'm working on it](https://github.com/Lyberta/cpp-unicode).

**Comment by expnkx** [comment]
Unicode is slightly different. Without going to files and sockets you don't really need to care about endianness so Unicode layer can sit on top of byte layer and sometimes even don't use it. That will require a different design. I'm working on it.


I do not think you understand what I am talking about. There is no reason to put them all together into a single class. However, they can be template. Just like std::deque<std::vector<std::vector<std::ofstream>>> works.

You can just add a Unicode Layer above your IO device and that will work.

Here is how my fast_io library works.
fast_io::ucs<fast_io::obuf,char32_t> ucs_obuf;
This will automatically encoding the stream without even notice it.

**Comment by Lyberta** [comment]
Yes, if you really need a byte device, you can put Unicode stream on top of such device. But in case of memory IO you really only care about code units and can even specify API in terms of Unicode scalar values so encoding form is hidden inside templates.

**Comment by expnkx** [comment]
> Yes, if you really need a byte device, you can put Unicode stream on top of such device. But in case of memory IO you really only care about code units and can even specify API in terms of Unicode scalar values so encoding form is hidden inside templates.


It is very simple when you do have such kind of devices. You can easily implement them.

https://bitbucket.org/ejsvifq_mabmip/fast_io/src/default/include/fast_io_freestanding_impl/ucs.h

For utf-8 stream, you do not care endianness.

**Comment by expnkx** [comment]
There also other requirements for IO library.

I am now working on freestanding IO since dynamic memory pool, floating-point and exception handling might be available in a freestanding environment.

**Comment by Lyberta** [comment]
> It is very simple when you do have such kind of devices. You can easily implement them.
> 
> https://bitbucket.org/ejsvifq_mabmip/fast_io/src/default/include/fast_io_freestanding_impl/ucs.h

Undocumented code dumps are not a good idea to make your message heard. I know that wrapping code unit to byte conversion is not difficult.

> For utf-8 stream, you do not care endianness.

You care about BOM.

**Comment by expnkx** [comment]
> > It is very simple when you do have such kind of devices. You can easily implement them.
> > https://bitbucket.org/ejsvifq_mabmip/fast_io/src/default/include/fast_io_freestanding_impl/ucs.h
> 
> Undocumented code dumps are not a good idea to make your message heard. I know that wrapping code unit to byte conversion is not difficult.
> 
> > For utf-8 stream, you do not care endianness.
> 
> You care about BOM.

For utf-8. No. It shouldn't be cared since utf-8 is endian free.

**Comment by Lyberta** [comment]
UTF-8 BOM is a thing and some users want it.

**Comment by expnkx** [comment]
> UTF-8 BOM is a thing and some users want it.

still does not change the fact they can be abstracted as a stream.

**Comment by Lyberta** [comment]
> still does not change the fact they can be abstracted as a stream.
The problem is that memory streams do not need BOM because everything is typed and is in native endianness. Byte streams sometimes need BOM for other programs. So this duality can be handled by different stream templates.

> The problem for me is that stdio.h and iostream are too slow. 10x slower(sometimes 100x) slower than doing it by hand.

Repeating this everywhere is not helpful. Most people know that it is slow.

> Unfortunately, network TS is still based on iostream. That is terrible.

Not really. Only `basic_socket_iostream` and you don't have to use it. You can safely use `basic_stream_socket`.

> EOF is, unfortunately, horrific stuff. Do not believe herbception would solve the problem. It won't.

I'm not sure if the wording specifies all corner cases at this point but `read_some` member function of input streams in my proposal simply returns `0` for EOF and `std::io::read` throws exception if it can't read all the bytes requested. This is elegant and works. Herbceptions will only make it faster.

**Comment by expnkx** [comment]
> > still does not change the fact they can be abstracted as a stream.
> 
> The problem is that memory streams do not need BOM because everything is typed and is in native endianness. Byte streams sometimes need BOM for other programs. So this duality can be handled by different stream templates.
> 
> > The problem for me is that stdio.h and iostream are too slow. 10x slower(sometimes 100x) slower than doing it by hand.
> 
> Repeating this everywhere is not helpful. Most people know that it is slow.
> 
> > Unfortunately, network TS is still based on iostream. That is terrible.
> 
> Not really. Only `basic_socket_iostream` and you don't have to use it. You can safely use `basic_stream_socket`.
> 
> > EOF is, unfortunately, horrific stuff. Do not believe herbception would solve the problem. It won't.
> 
> I'm not sure if the wording specifies all corner cases at this point but `read_some` member function of input streams in my proposal simply returns `0` for EOF and `std::io::read` throws exception if it can't read all the bytes requested. This is elegant and works. Herbceptions will only make it faster.

EOF is for functions like fgetc(). Not for functions like read

Herbceptions won't make it faster. It will make it at least 2 times slower. Herbception fails to replace both error code and exceptions.

**Comment by Lyberta** [comment]
> EOF is for functions like fgetc(). Not for functions like read. EOF is for getting single character from stream.

C API is broken. Don't use it.

> Herbceptions won't make it faster. It will make it at least 2 times slower. Herbception fails to replace both error code and exceptions.
> 
> The current table-based exception is the best we can get. His herbception sucks.

This is not the place to discuss this. I have seen only good stuff so far. Even if API is a bit involved.

> In my implementation, I do not use std::span. I do not think std::span won't lose performance here. I just use std::contiguous_iterator range (like algorithm) to do the abstraction.

I was considering supporting `std::ranges::contiguous_range` directly but now `std::span` got better constructors in Belfast so it just works.

With iterator pair you still have to call `std::to_address` to get the start and `std::distance` followed by `std::to_address` to get the end when interacting with OS files and sockets. So the amount of work is the same.

**Comment by expnkx** [comment]
> C API is broken. Don't use it.
> 
> >

However, C++ still has tons of APIs about this stuff. One example is std::char_traits.


With iterator pair you still have to call std::to_address to get the start and std::distance followed by std::to_address to get the end when interacting with OS files and sockets. So the amount of work is the same.

However, that would reduce the burden of both compiler and programmer, comparing to std::span.

**Comment by expnkx** [comment]
writes(out, std::span(vec.cbegin(),vec.cend()));
vs
writes(out, vec.cbegin(),vec.cend());

**Comment by expnkx** [comment]
BTW. contiguous_iterator range would work as normal C++ algorithms. That would make the parallel algorithm stuffs to applying to stream to be possible.

**Comment by expnkx** [comment]
Also, member functions are super broken for concepts. OOP is bad.

**Comment by Lyberta** [comment]
> However, C++ still has tons of APIs about this stuff. One example is std::char_traits.

Sure, for that reason my Unicode design doesn't use `std::char_traits`.

> However, that would reduce the burden of both compiler and programmer, comparing to std::span.

Care to elaborate?

> writes(out, std::span(vec.cbegin(),vec.cend()));
> vs
> writes(out, vec.cbegin(),vec.cend());

Uh, no. `std::vector` satisfies `std::ranges::contiguous_range` so `std::span` can be implicitly constructed from it.
```
writes(out, vec);
```

> Also, member functions are super broken for concepts. OOP is bad.

Only zealous inheritance + virtual functions where there is no need for them.

**Comment by expnkx** [comment]
> ```
> writes(out, vec);
> ```
> 
> 
> >

With contiguous_iterator, it is even better and works just like algorithm

**Comment by Lyberta** [comment]
> Do not trust the compiler to be guaranteed to optimize it.

`std::span<std::byte>` is a concrete class so optimizing it is the easiest.

**Comment by expnkx** [comment]
> > Do not trust the compiler to be guaranteed to optimize it.
> 
> `std::span<std::byte>` is a concrete class so optimizing it is the easiest.

You are using std::span<std::byte>??

**Comment by Lyberta** [comment]
> So you are using std::span<std::byte>??

Yes, have you even read the proposal?

> What if I want to writes std::vector<std::size_t>??

There are several ways to serialize it (size + objects, objects + sentinel value, something else) that is why you need to write serialization code for such case manually.

> That is the same problem with std::ofstream. People have to do explicit pointer cast to get it to work and often become the sources of buffer overflow.

Please provide more info. Remember that you can't serialize structs and classes directly, you need to serialize each "pod" member separately.

> That is exactly why I am using std::contiguous_iterator here. Automatically type punning.

What punning? Why?

**Comment by expnkx** [comment]
Yes, have you even read the proposal?

I did. However, I do not know whether that is your final goal because you said

writes(out, vec); with span. 

> There are several ways to serialize it (size + objects, objects + sentinel value, something else) that is why you need to write serialization code for such case manually.
> 
> >

I think that is the problem here. Serialization should be automatically with standard containers, or people will continue abusing JSON file for serialization instead of binary serialization.

**Comment by expnkx** [comment]
https://bitbucket.org/ejsvifq_mabmip/fast_io/src/default/examples/0008.binary_serialization/binary_serialization.cc

Here is my example of binary serialization.
write container. KISS

**Comment by expnkx** [comment]
https://bitbucket.org/ejsvifq_mabmip/fast_io/src/default/examples/0034.crypto/ctr_binary_serialization.cc

I can even do binary serialization with encryption/decryption by read/write

**Comment by Lyberta** [comment]
> I did. However, I do not know whether that is your final goal because you said
> 
> writes(out, vec); with span.

I just fixed your code and you didn't say the type of `vec` there so I assumed `std::vector<std::byte>`.

> I think that is the problem here. Serialization should be automatically with standard containers, or people will continue abusing JSON file for serialization instead of binary serialization.

No, no and no. I initially wrote the library to work with MIDI files, then expanded to WAV, QCVM, WebAssembly and other exotic formats. The goal of this proposal is that I give basic tools that can support any format and it is up to the user to write most of serialization code.

**Comment by expnkx** [comment]
> > I did. However, I do not know whether that is your final goal because you said
> > writes(out, vec); with span.
> 
> I just fixed your code and you didn't say the type of `vec` there so I assumed `std::vector<std::byte>`.
> 
> > I think that is the problem here. Serialization should be automatically with standard containers, or people will continue abusing JSON file for serialization instead of binary serialization.
> 
> No, no and no. I initially wrote the library to work with MIDI files, then expanded to WAV, QCVM, WebAssembly and other exotic formats. The goal of this proposal is that I give basic tools that can support any format and it is up to the user to write most of serialization code.

Then that is the problem. Currently we have different proposals for doing the exactly same thing.

Network TS + this +unicode + blah blah.

**Comment by expnkx** [comment]
It is unfortunately for C++, unlike JAVA, they have tons of stuffs that work with each other in their library while for C++ all these libraries talk really hard with each other.

**Comment by Lyberta** [comment]
> Then that is the problem. Currently, we have different proposals for doing the exact same thing in my opinion.
> 
> Network TS + this +unicode + blah blah.

You are welcome to contact relevant study groups. I develop this proposal with guidance of Niall Douglas and soon intend to sync APIs with Networking TS. I'm also in touch with Unicode study group.

> We do not even have a pipe in the standard library

Pipes can be added later on top of this proposal.

**Comment by expnkx** [comment]
> Niall Douglas

Niall Douglas is bad. His goal is to destroy C++.

**Comment by expnkx** [comment]
> > Then that is the problem. Currently, we have different proposals for doing the exact same thing in my opinion.
> > Network TS + this +unicode + blah blah.
> 
> You are welcome to contact relevant study groups. I develop this proposal with guidance of Niall Douglas and soon intend to sync APIs with Networking TS. I'm also in touch with Unicode study group.
> 
> > We do not even have a pipe in the standard library
> 
> Pipes can be added later on top of this proposal.

Another abomination stuff of iostream/stdio.h/charconv is that the base is a dynamic type instead of a template. That is a huge problem also why I said charconv is an abomination.

It has to do runtime checking to know whether the base is 10 or any other. That means It can NEVER be fast. Also, it requires to know the buffer size before read/write which means it is slow and can never be fast.

**Comment by expnkx** [comment]
https://bitbucket.org/ejsvifq_mabmip/fast_io/src/default/examples/0018.base/base.cc

Here is what I think how the base stuff should look like. Just provide the base in the template and the compiler knows what the base is. Then it can generate different tables for each base. Or it can NEVER be fast.

Unfortunately for charconv, it cannot achieve such stuff.

**Comment by Lyberta** [comment]
This proposal is only concerned with byte IO so far. Text handling is a different topic.

**Comment by expnkx** [comment]
> This proposal is only concerned with byte IO so far. Text handling is a different topic.

To be honest, text handling is a much larger problem for stdio.h/iostream than "byte IO" since these locale abominations are the real performance killer.

**Comment by Lyberta** [comment]
So? I'm not gonna fix entire C++ with one proposal.

**Comment by expnkx** [comment]
> So? I'm not gonna fix entire C++ with one proposal.

Then you realize that is the problem of C++? People just keep adding stuff and these stuff never work with each other correctly.

**Comment by expnkx** [comment]
I just did the performance benchmark again with latest GCC. You can see how slow fprintf and std::ofstream are. outputing 10 million unsigned integers take you 0.54 sec while the correct implementation only needs 0.07 sec on linux.
On windows, the gap is even larger. More than 20 times slower.

cqwrteur@DESKTOP-7H7UHQ9:~$ ./output_10M_size_t
std::FILE*:     0.54400720s
std::ofstream:  0.54688160s
std::ofstream with tricks:      0.35454000s
std::to_chars + ofstream rdbuf tricks:  0.15839360s
std::to_chars + obuf:   0.12315570s
obuf:   0.07055010s
obuf text:      0.14413950s
obuf ucs_view:  0.15338270s
dynamic obuf:   0.17027930s
obuf_mutex:     0.08181280s
fsync:  0.15912310s
speck128/128:   0.25258350s

**Comment by Lyberta** [comment]
We are getting offtopic here. Unicode IO will be a separate proposal. Non-Unicode - not sure.

**Comment by expnkx** [comment]
> We are getting offtopic here. Unicode IO will be a separate proposal. Non-Unicode - not sure.

Ok. Then why do you think binary serialization and print hello world is not part of this proposal?

**Comment by Lyberta** [comment]
This proposal has `std::io::read` and `std::io::write` for binary serialization.

"Hello world" is text so it requires Unicode layer which is too complex to put it in one proposal.

**Comment by expnkx** [comment]
> This proposal has `std::io::read` and `std::io::write` for binary serialization.
> 
> "Hello world" is text so it requires Unicode layer which is too complex to put it in one proposal.

Why Hello World requires the Unicode layer? I do not understand.

**Comment by expnkx** [comment]
I know you did a lot of work on the proposal.

However, it proves 3 isocpp study groups with hundreds of people did fewer jobs than 2 people me and my roommate (he only implements the cryptography part since he is a python programmer).

We have implemented binary serialization, Rty, socket, pipe, multi-precision, cryptography for 3 months while the entire isocpp SG with thousands of people do not even have a plan to fix the problem of outputting 10M integers? They do not even have a plan on how to do binary serialization or how to outputting 10M integers?

"This proposal has std::io::read and std::io::write for binary serialization."

Then how would you deal with std::vector<std::deque<std::vector<std::string>>> for your binary serialization?

**Comment by Lyberta** [comment]
> Why Hello World requires the Unicode layer? I do not understand.

Because anything that is not Unicode is just a bunch of random numbers that result in mojibake when you try to output them as text.

> Even C back in 1970s can printf hello world while we cannot do that now? There was no Unicode in the 1970s.

C could never do proper Unicode and anything that was produced by `printf` is in some implementation defined encoding that should never be used.

> Then how would you deal with `std::vector<std::deque<std::vector<std::string>>>` for your binary serialization?

You let the users decide. Stop advocating for One True Format. Not gonna happen.

**Comment by expnkx** [comment]
> > Why Hello World requires the Unicode layer? I do not understand.
> 
> Because anything that is not Unicode is just a bunch of random numbers that result in mojibake when you try to output them as text.

That is false since unicode/ucs/utf-8 does support ANSI as backward compatibility.

> 
> > Even C back in 1970s can printf hello world while we cannot do that now? There was no Unicode in the 1970s.
> 
> C could never do proper Unicode and anything that was produced by `printf` is in some implementation defined encoding that should never be used.

What? That is overengineering again.

> 
> > Then how would you deal with `std::vector<std::deque<std::vector<std::string>>>` for your binary serialization?
> 
> You let the users decide. Stop advocating for One True Format. Not gonna happen.

Let the users decide. Then these stuffs can never talk with each other and people will continue using JSON.

**Comment by Lyberta** [comment]
> That is false since unicode/ucs/utf-8 does support ANSI as backward compatibility.

What is ANSI? There was no support even for ASCII. `"Hello world"` literal may be in EBCDIC and it is on z/OS. Anything using `char` and narrow string literals is implementation defined i.e. mojibake.

> What? That is overengineering again. Even charconv does not care these stuffs.

Yes, it uses `char` which means it is broken and should never be used.

> Let the users decide. Then this stuff can never talk with each other and people will continue using JSON.

I was able to read MIDI, WAV, QuakeC, WebAssembly and other files just fine. JSON is also a good format. Much better than "plain text". Plain text is the worst.

> Standardization is to not let the users decide how they should use their language in their way or it just ends up with code no one could understand.

So just remove all the power and make C++ a toy language? Not gonna happen. If you have any problem understanding examples in this proposal, you can specify exact things you don't understand.

**Comment by expnkx** [comment]
> That is false since unicode/ucs/utf-8 does support ANSI as backward compatibility.
> 
> What is ANSI? There was no support even for ASCII. `"Hello world"` literal may be in EBCDIC and it is on z/OS. Anything using `char` and narrow string literals is implementation defined i.e. mojibake.

You do understand locale is a problem right? EBCDIC. Why? Why you cannot abstract your stream for templates for handling EBCDIC/unicode stuff? Why not assuming it is Unicode by default. I do not understand.

> 
> > What? That is overengineering again. Even charconv does not care these stuffs.
> 
> Yes, it uses `char` which means it is broken and should never be used.
> 
> > Let the users decide. Then this stuff can never talk with each other and people will continue using JSON.
> 
> I was able to read MIDI, WAV, QuakeC, WebAssembly and other files just fine. JSON is also a good format. Much better than "plain text". Plain text is the worst.
What do you mean of "plain text"? writes the entire content of std::vector<std::size_t> 

size+ content is "plain text"? you clearly do not understand what I am talking about.

> 
> > Standardization is to not let the users decide how they should use their language in their way or it just ends up with code no one could understand.
> 
> So just remove all the power and make C++ a toy language? Not gonna happen. If you have any problem understanding examples in this proposal, you can specify exact things you don't understand.

You sound like your proposal is not something toy.

**Comment by expnkx** [comment]
BTW. Your claim that C/C++ tries to support a wide range of hardware is false. On the contrary, many ACM papers said a lot of security problems are the hardware vendor tries to abstract their machine as C/C++ abstract machine.

**Comment by expnkx** [comment]
I show you an example. pointer. Back In 1970s PDP machine, you treat your memory as some kind of a bunch of contiguous address space.

However, today's machine is cache oblivion machine which means you have multiple lines of memory addresses you need to deal with, while in C/C++ you still use pointer to access them, Even they are multiple lines of different hardware.

**Comment by expnkx** [comment]
C Is Not a Low-level Language

https://queue.acm.org/detail.cfm?id=3212479

**Comment by expnkx** [comment]
I just did another same benchmark under windows. FILE* 27.5 times slower, std::ofstream 14 times slower than doing it by hand. Over 10x times slow is nowhere acceptable for IO. I do not believe byte io has such a performance gap compared to this.


D:\hg\fast_io\examples\build2>output_10M_size_t
std::FILE*:     2.20501300s
std::ofstream:  1.16499800s
std::ofstream with tricks:      1.03099900s
std::to_chars + ofstream rdbuf tricks:  0.82499900s
std::to_chars + obuf:   0.12199900s
obuf:   0.08299800s
obuf text:      0.23505600s
obuf ucs_view:  0.17200200s
dynamic obuf:   0.14499800s
obuf_mutex:     0.13999900s
fsync:  0.36406600s
speck128/128:   0.28299800s

**Comment by Lyberta** [comment]
> You do understand locale is a problem right?

It is much bigger problem. Nothing related to `char` is guaranteed to be Unicode.

> EBCDIC. Why?

It is still used on z/OS for narrow literals apparently.

> Why not assuming it is Unicode by default. I do not understand.

Because the standard explicitly says so.

> EBCDIC is dead. No one cares about it in 2019.

Tell that to IBM and Unicode Study Group. They care about EBCDIC a lot.

> The locale is bad. Unicode is the standard and C++ should accept the fact it is standard.

I don't own the standard. Otherwise I would remove a lot of nonsensical stuff from it and force UTF-8.

> > What do you mean of "plain text"? writes the entire content of std::vectorstd::size_t
> 
> size+ content is "plain text"? you clearly do not understand what I am talking about.
> 
> vector<size_t> vec(1,400);
> write_size(out,vec.size());
> writes(out,vec.cbegin(),vec.cend());
> 
> is plain text???

No it isn't. But now you've written `std::vector::size_type` directly to the stream which means your data can't be shared between 32 bit and 64 bit programs. Because the have different format of that type.

**Comment by expnkx** [comment]
No it isn't. But now you've written std::vector::size_type directly to the stream which means your data can't be shared between 32 bit and 64 bit programs. Because the have different format of that type.

Ha. the size type is flexible length. It can work on any 32bits and64bits program just like utf-8.

Also it has multiple different functions for writing data. print/write/json write in the future

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by Lyberta** [comment]
> Ha. the size type is flexible length. It can work on any 32bits and64bits program just like utf-8.

How is it encoded then?

**Comment by expnkx** [comment]
first bit tells the stream whether it stops.

100010000 10000000 00000000

equals to
00010000000000000000000

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by expnkx** [comment]
It had been a long term issue where write non trivial copyable type directly is undefined behavior. I think we can give it a reliable way to work

BTW outputing hello world or integer is endian free. EBCDIC does not matter

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by expnkx** [comment]
OH no

100010000 10000000 00000000 is wrong and not legal

flexible size

00000000 00000000 11000000

is
real size

1000000 0000000 0000000

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by Lyberta** [comment]
So what if I write a 64 bit size that can't fit into 32 bits and I read it on a 32 bit platform? Do you want to encode all integers this way?

Outputting `"hello world"` is not portable because bytes of `"hello world"` are not portable.

**Comment by expnkx** [comment]
Then the compiler is not compatible as well? No. You compile program and That program will work correctly

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by Lyberta** [comment]
`"hello world"` on US Windows will be in Windows-1252 while the same `"hello world"` on z/OS will be EBCDIC.

**Comment by expnkx** [comment]
Depends on your compiler by default. If your compiler is utf-8 then it is utf-8. If it is EBCDIC, it will be EBCDIC. I see nothing wrong here. Or C++17 standard library charconv is wrong as well

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by expnkx** [comment]
BTW. It can add another layer of template to DEAL with different encoding system without stupid C locale.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by Lyberta** [comment]
> Depends on your compiler by default.

Exactly. I don't want compiler to ruin my data. That's why I only use UTF-8. That is `char8_t`, `u8"hello world"` and the like.

Locales don't work with `char8_t` at all. They work with `char` so they are broken and I'm not interested in supporting them or anything that uses `char`.

**Comment by expnkx** [comment]
template can solve the problem

C locale clearly violates zero-overhead principle.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by Lyberta** [comment]
A proper Unicode layer for proper Unicode text at least. And I've said that it is on my TODO list but much later in a different proposal.

**Comment by expnkx** [comment]
All these are easily solvable issues. No technically hardness at all

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by Lyberta** [comment]
Then you've never implemented a constexpr version of Unicode character database.

**Comment by expnkx** [comment]
For me the most important stuff is How to output integer and floating point without any performance penalty.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by expnkx** [comment]
Why do you need table based stuff?

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by Lyberta** [comment]
How are you gonna implement a Unicode version of `std::isspace`?

**Comment by expnkx** [comment]
I don't see that is a problem since utf-8 encoding system was designed to avoid these problems in the beginning

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by Lyberta** [comment]
You need UCD to determine if something is whitespace. You need that to read text. Therefore, text layer requires a lot of work. A LOT. I have proven my point.

**Comment by expnkx** [comment]
Why? table based is slow

for ascii All spaces are either 32 or in the range 9-14

You just first do If for 32 and then do a minus with unsigned arithmetic are you are done. MUCH faster than table based

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by expnkx** [comment]
It is a myth table based stuff would be faster than If and switch. Accessing a random address is NEVER O(1)

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by Lyberta** [comment]
Take U+180E mongolian vowel separator.

It was not a whitespace Unicode 3.0, became whitespace in Unicode 4.0 and went back Unicode 6.3.

So `ucd.is_space(U'\u180E')` would return different values based on the version of `ucd` and therefore text parsing code will behave differently. [It is documented](https://codeblog.jonskeet.uk/2014/12/01/when-is-an-identifier-not-an-identifier-attack-of-the-mongolian-vowel-separator/).

I don't care about "just" ASCII. It's like 0.001% of Unicode. Your code is horribly wrong and naive. Stop claiming expertise over something you don't know about.

**Comment by expnkx** [comment]
Locale based stuff is horribly wrong
Assuming ascii is nothing wrong

Stop support stupid stuffs
Let IBM bankrupt

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by expnkx** [comment]
Like I said before. If outputing or inputing integer takes 20x slower than doing it by hand. It is a huge problem

Locale and entire CRT are clearly the huge violator of you don't pay If you don't use

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by expnkx** [comment]
BTW the iso c/c++ standard said very clearly

ONLY \n \t \r \v \f space are space. Other stuff ARE NOT space no matter what your encoding system is.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by expnkx** [comment]
C Locale is totally obsolete because nowadays character size is flexible instead of fixed size. Your table is exactly the same problem.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by Lyberta** [comment]
> BTW the iso c/c++ standard said very clearly ONLY \n \t \r \v \f space are space.

Only in the "C" locale which is not Unicode. I don't care much about "C" locale. Unicode is different.

> C Locale is totally obsolete because nowadays character size is flexible instead of fixed size.

Sure, Unicode extended grapheme clusters can be really huge. But my intuition says that I can iterate over scalar values, not grapheme clusters, when parsing text. That is to be determined in practice. This still requires having a UCD.

**Comment by expnkx** [comment]
Then just treat these 5 characters as space for eating spaces. Problem solved.

Please don't complicate the design.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by expnkx** [comment]
I would see a deprecation of C Runtime.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by Lyberta** [comment]
Unicode is complicated. There is no way around it. But it is also the only thing that actually solves encoding problem.

If you want ASCII only, you are free to write it, use it and propose it. But I don't think a lot of people need pure 7bit US-ASCII.

**Comment by expnkx** [comment]
I need utf-8 only. Just dealing ascii for isspace is good enough for 99% of people.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by expnkx** [comment]
These overcomplicated designs are exactly why i don't use C. Always damn stupid and slow.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by Lyberta** [comment]
> I need utf-8 only.

You do understand that UTF-8 encodes all ~1 million Unicode scalar values, right?

**Comment by expnkx** [comment]
UTF-8 and Unicode are two different systems.

Utf-8 is designed to be compatible with ascii. Treating  those 5 characters only as space is nothing wrong.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by expnkx** [comment]
Iterating utf-8 needs to be done with coroutine

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by Lyberta** [comment]
> UTF-8 and Unicode are two different systems.

UTF-8 is an encoding form (and encoding scheme) of Unicode scalar values. There are still ~1 million scalar values that can be encoded.

> Utf-8 is designed to be compatible with ascii.

Only if you use scalar values from ASCII range... which is a horribly wrong assumption for a standard library or any text processing system in 2019.

> Treating  those 5 characters only as space is nothing wrong.

It is very wrong.

> Iterating utf-8 needs to be done with coroutine

What the hell? What? Do you have any idea how UTF-8, code units, scalar values and grapheme clusters work?

**Comment by expnkx** [comment]
I don't see How it is terribly wrong for Treat 5 characters as space

utf-8 isn't unicode

Iterating utf-8, Yes only with coroutine.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by expnkx** [comment]
UTF-8 is an encoding form (and encoding scheme) of Unicode scalar values. There are still ~1 million scalar values that can be encoded.


still does not change the fact utf-8 and unicode are 2 different systems since utf-8 isn't the only encoding scheme or standard scheme of unicode

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by Lyberta** [comment]
> I don't see How it is terribly wrong for Treat 5 characters as space

Because there are other whitespace scalar values in Unicode. It is all determined by querying UCD. I'm tired of repeating that. I'm not gonna do that anymore.

> utf-8 isn't the only encoding scheme or standard scheme of unicode

So? The difference between encoding forms and schemes is how you iterate code units. Higher levels are the same.

**Comment by expnkx** [comment]
I don't care other space characters sorry and i am not the only person who doesn't care.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________
